pipeline_dp/dp_engine.py

```diff
@@ -207,14 +208,27 @@ def select_private_partitions(self, col,
         # to _way_ too many partitions.
         def sample_unique_elements_fn(pid_and_pks):
             pid, pks = pid_and_pks
-            unique_pks = set(pks)
-
-            sampled_elements = np.random.choice(np.array(list(unique_pks)),
-                                                max_partitions_contributed)
+            unique_pks = list(set(pks))
+            if len(unique_pks) <= max_partitions_contributed:
+                sampled_elements = unique_pks
+            else:
+                # np.random.choice makes casting of elements to numpy types
+                # which is undesirable by 2 reasons:
+                # 1. Apache Beam can not serialize numpy types.
+                # 2. It might lead for losing precision (e.g. arbitrary
+                # precision int is converted to int64).
+                # So np.random.choice should not be applied directly to
+                # 'unique_pks'. It is better to apply it to indices.
+                sampled_indices = np.random.choice(np.arange(len(unique_pks)),
+                                                   max_partitions_contributed,
+                                                   replace=False)
+
+                sampled_elements = [unique_pks[i] for i in sampled_indices]
```

Comment:
How about sampled_elements = unique_pks[sampled_indices]?

Reply:
using arrays for indexing works only for numpy arrays, unique_pks is a list and we don't want to convert it to numpy array because of types conversion :)

Reply:
Sounds good, thanks for clarifying!